src/combinator.rs

```diff
@@ -0,0 +1,304 @@
+/// ```
+/// use iom::{ pure, void, IO };
+///
+/// # fn main () -> std::io::Result<()> {
+/// // Passing this to `IO::run` would be invalid...
+/// let meow: IO<'_, &str> = pure("meow");
+/// // This *can* be executed, though.
+/// let action: IO<'_, ()> = void(meow);
+///
+/// IO::run(action)?;
+/// # Ok (()) }
+/// ```
+///
+/// `void(x)` is equivalent to `x.map(|_| ())` or `x.replace(())`.
+pub fn void <'a, T> (x: IO<'a, T>) -> IO<'a>
+where
+    T: 'a,
+{
+    x.replace(())
+}
+
+/// Return a function that, given another function, returns a chain of
+/// [`IO`] actions that are performed between performing `acquire` and
+/// `release`.
+///
+/// # Example
+///
+/// In the example below, `the_bracket` is a function that takes a
+/// function that takes `()` (since `println` returns `()`) and returns
+/// an instance of `IO`. Whatever that action is, the bracket we construct
+/// here guarantees that, before we run it, the message "start" will be
+/// printed, and after it has finished running, "finish" will be printed.
+/// ```
+/// use iom::{ println, bracket, IO };
+///
+/// # fn main () -> std::io::Result<()> {
+/// let the_bracket = bracket(println("start"), println("finish"));
+/// let action = the_bracket(|_| {
+///     println("busy...")
+/// });
+///
+/// let result = IO::capture(action)?;
+///
+/// assert_eq!(result, "start\nbusy...\nfinish\n");
+/// # Ok (()) }
+/// ```
+pub fn bracket <'a, I, T, F> (
+    acquire: IO<'a, I>,
+    release: IO<'a>,
+) -> impl FnOnce (F) -> IO<'a, T>
+where
+    F: FnOnce (I) -> IO<'a, T> + 'a,
+    T: 'a,
+    I: 'a,
+{
+    // Set up the state.
+    let IO (mut acquire) = acquire;
+    let IO (mut release) = release;
+
+    |f: F| IO::from_fn(move |Context { i, o, e }| {
+
+        // Acquire the resource.
+        let x = acquire.perform(Context { i, o, e })?;
+        // Apply the function to the resource.
+        let IO (mut a) = f(x);
+
+        // Perform the effect but do not return early if it fails
+        // since we still need to do cleanup.
+        let result = a.perform(Context { i, o, e });
+        // Perform cleanup.
+        release.perform(Context { i, o, e })?;
+
+        // Return the result of the action resulting from
+        // the application of `f` to the resource.
+        result
+
+    })
+
+}
+
+/// Various combinators on the IO monad.
+impl<'a, T> IO<'a, T>
+where
+    T: 'a,
+{
+
+    /// Perform a monadic binding operation. For the IO monad, this
+    /// means performing `self`, calling `f` on the result and then
+    /// performing the [`IO`] actions returned from `f`.
+    ///
+    /// You can think of this as the [`IO`] version of [`Option`]'s
+    /// [`and_then`](Option::and_then) function, except fully lazy.
+    ///
+    /// If we simplify the signatures somewhat, the similarities are
+    /// clear(er) to see:
+    ///
+    /// ```text
+    /// fn and_then <F, U> (self: Option<T>, f: F) -> Option<U>
+    /// where
+    ///     F: FnOnce (T) -> Option<U>;
```

Comment:
Maybe note the place where `and_then` and `bind` differ from `map`
```suggestion
    ///     F: FnOnce (T) -> Option<U>;
    ///     //                          ------------ Note the return value here
```